### PR TITLE
feat(ui): frontend implementation for avatar uploading in profile

### PR DESCRIPTION
**Pull request by wwayne**
Close TAB-541

Screen record: https://jam.dev/c/4558b9c5-9a28-44b3-8e63-7cd290333547

### CODE REVIEW

File: ee/tabby-webserver/development/Caddyfile [hunk 1]
```diff
@@ -8,6 +8,7 @@
 		path /hub
 		path /repositories/*
 		path /oauth/*
+    path /avatar/*
```
Comment (wsxiaoys) [inline ee/tabby-webserver/development/Caddyfile:11]: please run caddy format

Comment (wwayne) [reply ee/tabby-webserver/development/Caddyfile:11]: did `caddy fmt --overwrite`

**Review by wsxiaoys — APPROVED** [review]
LG from UX perspective

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx [hunk 2]
```diff
@@ -1,17 +1,113 @@
...
+        <Button
+          type="submit"
+          disabled={!uploadedImgString || isSubmitting}
+          onClick={onUploadAvatar}
+          className="w-40"
+        >
+          {isSubmitting && (
+            <IconSpinner className="mr-2 h-4 w-4 animate-spin" />
+          )}
+          Update avatar
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx:104]:
```suggestion
          Save Changes
```

File: ee/tabby-ui/app/(dashboard)/profile/components/change-password.tsx [hunk 3]
```diff
@@ -132,7 +132,7 @@ const ChangePasswordForm: React.FC<ChangePasswordFormProps> = ({
         <FormMessage />
         <Separator />
         <div className="flex">
-          <Button type="submit" disabled={isSubmitting}>
+          <Button type="submit" disabled={isSubmitting} className="w-40">
             {isSubmitting && (
               <IconSpinner className="mr-2 h-4 w-4 animate-spin" />
             )}
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(dashboard)/profile/components/change-password.tsx:138]: `Save Changes`

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx [hunk 4]
```diff
@@ -1,17 +1,113 @@
...
+      const response = await uploadUserAvatar({
+        avatarBase64,
+        id: data.me.id
+      })
+      if (response?.error) throw response.error
+      if (response?.data?.uploadUserAvatarBase64 === false)
+        throw new Error('Upload failed')
+      await delay(1000)
+      mutateAvatar(data.me.id)
+      toast.success('Avatar uploaded successfully.')
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx:57]:
```suggestion
      toast.success('Successfully updated your profile picture!')
```

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx [hunk 5]
```diff
@@ -1,17 +1,113 @@
...
+      const mimeHeaderMatcher = new RegExp('^data:image/.+;base64,')
+      const avatarBase64 = uploadedImgString.replace(mimeHeaderMatcher, '')
+
+      const response = await uploadUserAvatar({
+        avatarBase64,
+        id: data.me.id
+      })
+      if (response?.error) throw response.error
+      if (response?.data?.uploadUserAvatarBase64 === false)
+        throw new Error('Upload failed')
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx:54]:
```suggestion
    // Handle error with toast.error
```

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx [hunk 6]
```diff
@@ -1,17 +1,113 @@
...
+
+      reader.readAsDataURL(file)
+    }
+  }
+
+  const onUploadAvatar = async () => {
+    setIsSubmitting(true)
+    try {
+      const mimeHeaderMatcher = new RegExp('^data:image/.+;base64,')
+      const avatarBase64 = uploadedImgString.replace(mimeHeaderMatcher, '')
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx:46]: should be better library / approach to handle the parsing part?

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx [hunk 7]
```diff
@@ -1,17 +1,113 @@
...
+          type="file"
+          accept="image/*"
+          className="hidden"
+          onChange={onPreviewAvatar}
+        />
+        {uploadedImgString && (
+          <img
+            src={uploadedImgString}
+            className="absolute left-0 top-0 z-10 h-16 w-16 rounded-full border object-cover"
+            alt="Upload image preview"
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx:86]:
```suggestion
            alt="avatar to be uploaded"
```

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx [hunk 8]
```diff
@@ -1,17 +1,113 @@
...
+          className="w-40"
+        >
+          {isSubmitting && (
+            <IconSpinner className="mr-2 h-4 w-4 animate-spin" />
+          )}
+          Update avatar
+        </Button>
+
+        <p className="mt-1.5 text-sm text-muted-foreground">
+          Square image recommended. Accept file types: .png, .jpg.
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx:108]:
```suggestion
          Square image recommended. Accepted file types: .png, .jpg. Max file size: 500KB.
```

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx (hunk 6 above)
Comment (wwayne) [reply ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx:46]: Did find good recommendation from gpt
its method is more straight forward
```js
// Get the result from the FileReader
const base64Data = reader.result;

// Split the string on the comma to remove the MIME type information
const base64String = base64Data.split(',')[1];
```
I would take this `split` method

File: ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx (hunk 5 above)
Comment (wwayne) [reply ee/tabby-ui/app/(dashboard)/profile/components/avatar.tsx]:
```js
const uploadUserAvatar = useMutation(uploadUserAvatarMutation, {
    onError(err) {
      toast.error(err.message)
    }
  })

const onUploadAvatar = async () => {
  setIsSubmitting(true)

  const response = await uploadUserAvatar({
    avatarBase64: uploadedImgString.split(',')[1],
    id: data.me.id
  })

  if (response?.data?.uploadUserAvatarBase64 === true) {
    await delay(1000)
    mutateAvatar(data.me.id)
    toast.success('Successfully updated your profile picture!')

    await delay(200)
    setUploadedImgString('')
  }
  
  setIsSubmitting(false)
}
```

**Review by liangfung — APPROVED** [review]

**Pull request merged**